Branch/save

Martin0071 opened the pull request:
Added save functionality

Martin0071 committed:
Working on save method

veselin-petkov committed:
working save for a student with name only

penzata commented on src/main/java/org/example/persistence/ormanager/ORManagerImplTest.java:
```diff
@@ -0,0 +1,15 @@
+package org.example.persistence.ormanager;
```
It is best all the test files to be in the test directory.

penzata commented on src/main/java/org/example/persistence/ormanager/ORManager.java:
```diff
@@ -49,4 +50,6 @@ static ORManager withPropertiesFrom(String filename) throws SQLException {
     static ORManager withDataSource(DataSource dataSource) {
         return new ORManagerImpl(dataSource);
     }
+    Student save(Student student);
```
Wouldn't it be better to use generics for the CRUD methods in the interface.

penzata commented on src/main/java/org/example/persistence/ormanager/ORManagerImpl.java:
```diff
@@ -28,4 +28,22 @@ static DataSource createDataSource(String url, String user, String password, Map
     }
+    @Override
+    public Student save(Student student){
+        Connection connection;
```
We are going to use connection and statement in all of our methods. Maybe consider making them fields.

penzata commented on src/main/java/org/example/persistence/ormanager/ORManagerImpl.java:
```diff
@@ -28,4 +28,22 @@ static DataSource createDataSource(String url, String user, String password, Map
+    @Override
+    public Student save(Student student){
+        Connection connection;
+        try {
```
Maybe try using try-with-resources here for the connection and the statement.

penzata commented on src/main/java/org/example/persistence/ormanager/ORManagerImpl.java:
```diff
@@ -28,4 +28,22 @@ static DataSource createDataSource(String url, String user, String password, Map
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+        return student;
```
As of this, we are returning the same object (with null Id state) that was passed in the method initially without any change.
Consider getting the auto-generated ID for this object from the database, set it, and then return the saved object with the appropriate id.

penzata commented on src/main/java/org/example/persistence/ormanager/ORManagerImpl.java:
```diff
@@ -28,4 +28,22 @@ static DataSource createDataSource(String url, String user, String password, Map
+        try {
+           connection = dataSource.getConnection();
+            Statement statement = connection.createStatement();
+            statement.execute("CREATE TABLE IF NOT EXISTS STUDENTS (first_name VARCHAR(255)) ;");
```
About this row and two rows below (row 39): it is always considered good practice to make the sql statements (or the most often used ones) constants and pass a constant in the methods, not the whole string itself.

And about the table creation: right now you are creating a table with just one column. Try checking out the table schema file in the resources directory.

penzata commented on src/main/java/org/example/persistence/ormanager/ORManagerImpl.java:
```diff
@@ -28,4 +28,22 @@ static DataSource createDataSource(String url, String user, String password, Map
+           PreparedStatement ps =  connection.prepareStatement("INSERT INTO STUDENTS (first_name) values(?)");
+           ps.setString(1, student.getFirstName());
+           int rows = ps.executeUpdate();
+            System.out.println(rows + " rows affected.");
```
Please, don't commit any System.out print lines. If you want to see your current result use Logger for debugging.

penzata commented on src/main/java/org/example/persistence/ormanager/ORManagerImpl.java:
```diff
@@ -28,4 +28,22 @@ static DataSource createDataSource(String url, String user, String password, Map
+            System.out.println(rows + " rows affected.");
+
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
```
Consider handling the exception, not throwing a new one.
Here's a good information about handling that you can use: https://docs.oracle.com/javase/tutorial/jdbc/basics/sqlexception.html

penzata commented on src/main/java/org/example/persistence/ormanager/ORManagerImpl.java:
```diff
@@ -1,14 +1,14 @@
 import java.util.Map;
 
-public class ORManagerImpl implements ORManager{
+public class ORManagerImpl implements org.example.persistence.ormanager.ORManager {
```
Why using the whole name of the ORManager interface?

penzata reviewed (changes requested):
Please, consider all the comments.

Martin0071 committed:
Implemented suggested changes

Added logger instead of prints, created constants for the sql queries, made save work with generics (work in progress), made connection and statement fields.

Martin0071 committed:
Removed target folder

Martin0071 committed:
Refactoring

Martin0071 committed:
We are stumped

veselin-petkov committed:
save method

veselin-petkov committed:
Merge remote-tracking branch 'origin/main' into branch/save

# Conflicts:
#	.gitignore
#	pom.xml
#	src/database/ORManager.mv.db
#	src/main/java/org/example/demo/Main.java
#	src/main/java/org/example/persistence/ormanager/ORManager.java
#	src/main/java/org/example/persistence/ormanager/ORManagerImpl.java

veselin-petkov committed:
save method

veselin-petkov committed:
h2.properties to be from a file
